File: loadercomplex-forge-18/src/main/java/com/enderzombi102/loadercomplex/forge18/impl/entity/ForgeEntity.java
package com.enderzombi102.loadercomplex.forge18.impl.entity;

import com.enderzombi102.loadercomplex.api.entity.Entity;
import com.enderzombi102.loadercomplex.api.entity.ItemEntity;
import com.enderzombi102.loadercomplex.api.entity.LivingEntity;
import com.enderzombi102.loadercomplex.api.utils.Position;
import com.enderzombi102.loadercomplex.api.world.World;
import com.enderzombi102.loadercomplex.forge18.impl.utils.BlockUtils;
import com.enderzombi102.loadercomplex.forge18.impl.world.ForgeWorld;
import net.minecraft.text.Text;

import java.util.UUID;

public class ForgeEntity implements Entity {
	private final net.minecraft.entity.Entity backingEntity;

	public ForgeEntity(net.minecraft.entity.Entity backingEntity) {
		this.backingEntity = backingEntity;
	}

	@Override
	public String getDisplayName() {
		return Text.Serializer.toJson( this.backingEntity.getName() );
	}

	@Override
	public String getName() {
		return this.backingEntity.getEntityName();
	}

	@Override
	public UUID getUuid() {
		return this.backingEntity.getUuid();
	}

	@Override
	public boolean isLivingEntity() {
		return this.backingEntity instanceof net.minecraft.entity.LivingEntity;
	}

	@Override
	public LivingEntity asLivingEntity() {
		return new ForgeLivingEntity( (net.minecraft.entity.LivingEntity) this.backingEntity );
	}

	@Override
	public boolean isItem() {
		return this.backingEntity instanceof net.minecraft.entity.ItemEntity;
	}

	@Override
	public ItemEntity asItem() {
		return new ForgeItemEntity( (net.minecraft.entity.ItemEntity) this.backingEntity);
	}

	@Override
	public void kill() {
		this.backingEntity.remove(net.minecraft.entity.Entity.RemovalReason.KILLED);
	}

	@Override
	public boolean isDead() {
		return this.backingEntity.isRemoved();
	}

	@Override
	public Position getPosition() {
		return BlockUtils.toPosition( this.backingEntity.getBlockPos() );
	}

	@Override
	public void setPosition(Position pos) {
		this.backingEntity.setPosition( pos.x, pos.y, pos.z );
	}

	@Override
	public float getPitch() {
		return this.backingEntity.getPitch();
	}

	@Override
	public float getYaw() {
		return this.backingEntity.getYaw();
	}

	@Override
	public World getWorld() {
		return new ForgeWorld( this.backingEntity.world );
	}

	@Override
	public Object getObject() {
		return this.backingEntity;
	}
}

File: loadercomplex-forge/src/ver182/kotlin/com/enderzombi102/loadercomplex/forge182/impl/entity/ForgeEntity.kt
package com.enderzombi102.loadercomplex.forge182.impl.entity

import com.enderzombi102.loadercomplex.api.math.Vec3d
import com.enderzombi102.loadercomplex.api.minecraft.world.World
import com.enderzombi102.loadercomplex.forge182.impl.utils.toLC
import com.enderzombi102.loadercomplex.forge182.impl.world.ForgeWorld
import net.minecraft.entity.Entity
import net.minecraft.entity.ItemEntity
import net.minecraft.entity.LivingEntity
import net.minecraft.text.Text
import java.util.*

open class ForgeEntity(private val backingEntity: Entity) :
	com.enderzombi102.loadercomplex.api.minecraft.entity.Entity {
	override fun getDisplayName(): String {
		return Text.Serializer.toJson(backingEntity.name)
	}

	override fun getName(): String {
		return backingEntity.entityName
	}

	override fun getUuid(): UUID {
		return backingEntity.uuid
	}

	override fun isLivingEntity(): Boolean {
		return backingEntity is LivingEntity
	}

	override fun asLivingEntity(): com.enderzombi102.loadercomplex.api.minecraft.entity.LivingEntity {
		return ForgeLivingEntity(backingEntity as LivingEntity)
	}

	override fun isItem(): Boolean {
		return backingEntity is ItemEntity
	}

	override fun asItem(): com.enderzombi102.loadercomplex.api.minecraft.entity.ItemEntity {
		return ForgeItemEntity(backingEntity as ItemEntity)
	}

	override fun kill() {
		backingEntity.remove(Entity.RemovalReason.KILLED)
	}

	override fun isDead(): Boolean {
		return backingEntity.isRemoved
	}

	override fun getPosition(): Vec3d {
		return backingEntity.blockPos.toLC().toVec3d()
	}

	override fun setPosition(pos: Vec3d) {
		backingEntity.setPosition(pos.x, pos.y, pos.z)
	}

	override fun getPitch(): Float {
		return backingEntity.pitch
	}

	override fun getYaw(): Float {
		return backingEntity.yaw
	}

	override fun getWorld(): World {
		return ForgeWorld(backingEntity.world)
	}

	override fun getObject(): Any {
		return this.backingEntity
	}
}
